### PR TITLE
Release v3.1.0

### PR DESCRIPTION
**Pull request by eloch216**
We have accumulated quite a few changes since our last release. Some of these are important to users, especially the one regarding warnings in module tests. I think it's time to make a new release.

Here, I've updated `cran-comments.md` based on the output from `R CMD check` on my own machine. It's possible that other notes will pop up on other operating systems, and we can update this document accordingly.

When this PR is completed, I think we should also consider submitting it to CRAN. I think we have done our due diligence in addressing `R CMD check` warnings. The contents of `cran-comments.md` are intended to be sent to CRAN when submitting the package for consideration.

More details about CRAN submission can be found here (I haven't seen this site before, but it seems like it might be helpful): https://r-pkgs.org/release.html. We can address any smaller last-minute changes as part of this PR.

### CODE REVIEW

**Comment by justinmcgrath** [comment]
It would be good to update the year in LICENSE and LICENSE.md. I was going to edit the files in this branch, but wasn't sure that was the workflow we agreed on.

**Comment by justinmcgrath** [comment]
This one is annoying: "CRAN checks for https URLs and will not allow any http links." The links will need to use https.

**Comment by eloch216** [comment]
> It would be good to update the year in LICENSE and LICENSE.md. I was going to edit the files in this branch, but wasn't sure that was the workflow we agreed on.

Personally, I think I am okay with making small changes in a release branch. Especially something like updating the year, which is related to making a release. Side note: I have always been unsure whether the year in the license should be the current year, or the year when the project started.

**Comment by justinmcgrath** [comment]
> Side note: I have always been unsure whether the year in the license should be the current year, or the year when the project started.

I don't know. I don't see why it would matter, but the site you linked to said CRAN would complain about it if it wasn't the current year.

**Comment by eloch216** [comment]
I'm looking through the check results from the Github actions. The package size seems to be different on each operating system, so I'm not sure what we should write about that in `cran-comments`. Most of the differences are small, but the `libs` sub-directory on Ubuntu is 25 MB, which is much larger than on Windows or MacOS where it is 1-4 MB.

**Comment by eloch216** [comment]
> This one is annoying: "CRAN checks for https URLs and will not allow any http links." The links will need to use https.

I started looking for them. Here's what I've found:
- Line 45 of `man\lightME.Rd`. This is a dead link anyway.
- Line 59 of `man\soil_parameters.Rd`. This is an `http` embedded in a wayback machine link.
- Lines 1247, 1252, 1257 of `vignettes\a_practical_guide_to_biocro.Rnw`

In a sane world, any links in the C++ code or in comments shouldn't matter.

I can address these ones that I found.

**Comment by gsrohde** [comment]
> > This one is annoying: "CRAN checks for https URLs and will not allow any http links." The links will need to use https.
> 
> I started looking for them.

Just out of curiosity, how are you finding these?  Are you just using grep or some Windows equivalent, or is there an R package that finds these?

By the way, [W3C's link checker](https://validator.w3.org/checklink) might come in handy at some point, even if it doesn't check exactly the same things CRAN might check.  I tried it with the URL for the public documentation (https://biocro.github.io/BioCro-documentation/latest/pkgdown/).  The only broken links it found, I think, were ones this PR fixes.  (Being able to use this tool on the documentation for a proposed release might be one argument for bringing back deployment of generated docs for other than the main branch.)

**Comment by eloch216** [comment]
> > > This one is annoying: "CRAN checks for https URLs and will not allow any http links." The links will need to use https.
> > 
> > 
> > I started looking for them.
> 
> Just out of curiosity, how are you finding these? Are you just using grep or some Windows equivalent, or is there an R package that finds these?

I just searched for "http" (as a case-sensitive whole word) in the BioCro repository using the search feature in `Notepad++`, and then manually looked to see which search results seemed legit. Probably not even as fancy as using `grep`.

**Review by gsrohde — COMMENTED** [review]
FWIW, I get a few test errors, even though there are none in the GitHub Actions tab:
```
* checking tests ... [66s/67s] ERROR
  Running ‘testthat.R’ [66s/67s]
Running the tests in ‘tests/testthat.R’ failed.
Last 13 lines of output:
  [ FAIL 2 | WARN 3 | SKIP 1 | PASS 5091 ]
  
  ══ Skipped tests (1) ═══════════════════════════════════════════════════════════
  • Skipping some tests of partial_run_biocro pending changes (1):
    'test.PartialApplication.R:79:5'
  
  ══ Failed tests ════════════════════════════════════════════════════════════════
  ── Failure ('test.NumericalIntegratorSettings.R:324:9'): The ODEINT Rosenbrock numerical ode_solver output is more accurate for smaller tolerances ──
  abs(rsnbrk_result_better_rel - best_result) is not strictly less than abs(rsnbrk_result_bad_settings - best_result). Difference: 0
  ── Failure ('test.NumericalIntegratorSettings.R:329:9'): The ODEINT Rosenbrock numerical ode_solver output is more accurate for smaller tolerances ──
  abs(rsnbrk_result_better_abs - best_result) is not strictly less than abs(rsnbrk_result_bad_settings - best_result). Difference: 0
  
  [ FAIL 2 | WARN 3 | SKIP 1 | PASS 5091 ]
  Error: Test failures
  Execution halted
```

Also, I get several dozen warnings coming from the `.Rd` files.  Here is a sample:
```
add_time_to_weather_data.Rd:36:1: Warning: <table> lacks "summary" attribute
annualDB.Rd:4:1: Warning: <link> inserting "type" attribute
annualDB.Rd:12:1: Warning: <script> proprietary attribute "onload"
```
All, or most all of the several dozen warnings are one of these three.

I don't recall seeing so many of these errors before.  A hunch is that they represent a move to make the resulting HTML pages more accessibility compliant.  I was using R 4.2.2.

**Comment by justinmcgrath** [comment]
> Just out of curiosity, how are you finding these? Are you just using grep or some Windows equivalent, or is there an R package that finds these?

I used vimgrep in vim which is mostly like grep.

>  The only broken links it found, I think, were ones this PR fixes.

Is there an automated way to check for broken links? I found one in the framework files and then quickly stopped checking because it was slow to do.

File: DESCRIPTION [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Package: BioCro
-Version: 3.0.2
+Version: 3.1.0
 Date: 2023-10-30
```
[the remark this answers is not in the document]
Comment (eloch216) [reply DESCRIPTION]: I'll be an optimist and hope we get this submitted within a month. I'll go ahead and change it.

**Comment by gsrohde** [comment]
> Is there an automated way to check for broken links? I found one in the framework files and then quickly stopped checking because it was slow to do.

The W3C checker I mentioned the the comment above (https://github.com/biocro/biocro/pull/88#issuecomment-1967579237) will check links in Web pages on public sites.  It's not clear to me exactly which links it checks when you click the option to check links recursively.  I tried putting in the address for BioCro on GitHub, https://github.com/biocro/biocro, figuring that might be a way to check links in files that don't normally form part of a Web site, since ultimately, on GitHub, you can reach any file in your repository via links.  But this also starts checking links in issues and pull requests, so it doesn't seem too useful unless there is a way to restrict the set of files looked at.

There are probably scripts out there that have a wider range of features and might even work on just sets of files on your own machine (I once wrote such as script myself) but I haven't tried looking.

**Comment by gsrohde** [comment]
> Most of the differences are small, but the `libs` sub-directory on Ubuntu is 25 MB, which is much larger than on Windows or MacOS where it is 1-4 MB.

The only thing in my `library/BioCro/libs` directory is `BioCro.so`, which is about 1.9MB.  The whole of `library/BioCro` contains about 5.7MB.  On Ubuntu, is there anything else in `libs` besides the BioCro lib file?

**Comment by gsrohde** [comment]
> > On Ubuntu, is there anything else in `libs` besides the BioCro lib file?
> 
> It's only that file, but it's 26 MB.

Do you have a `~/.R/Makevars` file or a R_MAKEVARS_SITE file?  I tried putting `CXX11FLAGS = -Wall -g -O0 $(LTO)` in my `~/.R/Makevars` file to override the `-O2` setting in `${R_HOME}/etc/Makeconf`.  With the added debugging symbols, that blew up the size of `BioCro.so` five times, to 9.5MB, though that's still a long ways from 26 MB.

**Comment by justinmcgrath** [comment]
> Do you have a `~/.R/Makevars` file or a R_MAKEVARS_SITE file?
I don't, but it does compile with the `-g` flag. If I create `~/.R/Makevars` with `CXX11FLAGS = -Wall -O2 $(LTO)` then BioCro.so is only 2 MB. I'm not sure where it's getting the `-g` flag from. Give the similar file sizes, it seems like the test machine is doing the same thing, but I couldn't find the compilation options. On my Windows machine, it doesn't use `-g`.

Maybe I have customized a Makefile somewhere to add `-g` on Ubuntu, but I don't recall doing that and can't find anything with it, so it appears to be the default. However, I can't find anyone that mentions this. If it is the default, then I think we're largely in line with other package sizes since the sizes on Windows and Mac are pretty small.

**Comment by gsrohde** [comment]
> > Do you have a `~/.R/Makevars` file or a R_MAKEVARS_SITE file?
> 
> I don't, 

The standard location for the Makevars site file is _R_HOME_/etc/_R_ARCH_/Makevars.site, so you should check for the existence of that file as well.

You could probably check whether a non-standard location for a personal or site-wide Makevars file has been set by seeing if `Sys.getenv("R_MAKEVARS_USER")` or `Sys.getenv("R_MAKEVARS_SITE")` returns a non-empty value.

There's more information about all of this at https://cran.r-project.org/doc/manuals/r-release/R-admin.html#Customizing-package-compilation.

> but it does compile with the `-g` flag. If I create `~/.R/Makevars` with `CXX11FLAGS = -Wall -O2 $(LTO)` then BioCro.so is only 2 MB. I'm not sure where it's getting the `-g` flag from. Give the similar file sizes, it seems like the test machine is doing the same thing, but I couldn't find the compilation options. On my Windows machine, it doesn't use `-g`.
> 
> Maybe I have customized a Makefile somewhere to add `-g` on Ubuntu, but I don't recall doing that and can't find anything with it, so it appears to be the default. However, I can't find anyone that mentions this. If it is the default, then I think we're largely in line with other package sizes since the sizes on Windows and Mac are pretty small.

On my Mac, _R_HOME_/etc/R_ARCH/Makeconf includes the line `CXX11FLAGS = -Wall -g -O2 $(LTO)`, and presumably this is the case on Ubuntu as well and this is where the `-g` comes from.  It surprised me a little that R uses the `-g` by default: removing it reduced the library size by about a quarter.  Still, the optimization level flag had a much larger effect on library size.

As an experiment, you could see what a typical compile command looks like during a package-install, both with and without the `~/.R/Makevars` overrides.  You could also try putting `CXX11FLAGS = -Wall -g -O0 $(LTO)` in your `~/.R/Makevars` to see if the library file size goes back up to 26MB.  Or, you could remove the setting of CXX11FLAGS from  `~/.R/Makevars` and instead replace it with `$(error $(CXX11FLAGS))` to get Make to print out the default value for `CXX11FLAGS`.

**Comment by justinmcgrath** [comment]
> As an experiment, you could see what a typical compile command looks like during a package-install, both with and without the `~/.R/Makevars` overrides. You could also try putting `CXX11FLAGS = -Wall -g -O0 $(LTO)` in your `~/.R/Makevars` to see if the library file size goes back up to 26MB.

-g -O0: 13 MB
-g -O2: 26 MB
-O2: 1.5 MB
-O0: 4.3 MB.

There's some interactive effect since -g -O2 is larger than -g -O0, but -O0 is larger than -O2.

I agree that it's odd to use -g by default. My understanding was that the debugging info could only be used in a debugger, so I don't see how it's useful. Nonetheless, we don't look unusual compared to other packages, so the size shouldn't be an issue with CRAN.

**Comment by justinmcgrath** [comment]
This is only tangentially related, but I found it out while looking at this. I have finally found how to build projects using parallel make.
On UNIX-like OSes you can use `MAKEFLAGS=-j15 R CMD INSTALL`, replacing 15 with what you'd like. One fewer than the number of cores seems to work well.

On Windows you can't temporarily set a variable. So you can do `set MAKEFLAGS=-j8` then `R CMD INSTALL`, and keep the variable set or unset it later.

You can pair this with ccache using `MAKEFLAGS=-j15 CC=ccache`.

This is pretty simply. I don't know how I didn't find it earlier.

**Comment by eloch216** [comment]
Just wanted to re-iterate that differences in package size can be seen in the GitHub action results. So it's not something unique to Justin's Ubuntu setup.

**Comment by eloch216** [comment]
> Also, I get several dozen warnings coming from the `.Rd` files. Here is a sample:

Seems like these might have something to do with `tidy` being on your path (whatever that is): https://groups.google.com/g/r-sig-mac/c/7u_ivEj4zhM?pli=1

**Comment by gsrohde** [comment]
> > Also, I get several dozen warnings coming from the `.Rd` files. Here is a sample:
> 
> Seems like these might have something to do with `tidy` being on your path (whatever that is): https://groups.google.com/g/r-sig-mac/c/7u_ivEj4zhM?pli=1

(From the Tidy web page (https://www.html-tidy.org/): "Tidy is a console application for macOS, Linux, Windows, UNIX, and more. It corrects and cleans up HTML and XML documents by fixing markup errors and upgrading legacy code to modern standards.")

Interesting discussion, and news to me.  But a bit baffling.  This comment in one of the posts linked to in the discussion (https://stat.ethz.ch/pipermail/r-package-devel/2022q1/007806.html) says it all: "This seems to be the new check for HTML documentation validity as part of accessibility enhancements in R. As far as I'm aware, HTML5 doesn't _require_ tables to have summaries, **_nor does Rd provide a way to specify them_** [bold emphasis mine]."  As I had guessed, this new checking has to do with Web page accessibility.  I'm all for accessibility, but why flag errors that R's Rd2HTML process itself generates, i.e., that can't be fixed by rewriting the Rd files? (Or can they?)

As one of the discussion comments notes, this new part of checking can be circumvented by setting an environment variable.  From the R 4.2.1 changes log: "Setting environment variable _R_CHECK_RD_VALIDATE_RD2HTML_ to a false value will override R CMD check --as-cran and turn off HTML validation. This provides a way to circumvent a problematic tidy."

This issue about "problematic tidy" and the talk about the old 2006 version of Tidy on macOS in the discussion you cited strikes me as being a bit of a red herring.  Even after upgrading my Tidy installation, I _still_ get lots of warnings about the Rd files, if slightly different ones, and as far as I know, these aren't fixable by revising the Rd files, though possibly I am wrong about this.

**Comment by eloch216** [comment]
Looking through the checklists at https://r-pkgs.org/release.html and https://github.com/DavisVaughan/extrachecks, I think we are in good shape. The only thing I'm unsure about is a `noSuggests` test, but it sounds like CRAN doesn't usually do this.

I think we are ready to release this new version, and then submit to CRAN. @gsrohde and @justinmcgrath do you agree?

**Comment by justinmcgrath** [comment]
I think it looks good. Thanks.

File: NEWS.md [hunk 2]
```diff
@@ -1,21 +1,137 @@
...
+      the --as-cran option.  Formerly, this was always used.
+
+    - The user can choose whether and when to throw an error on R CMD
+      check failures.  Formerly, an error was thrown whenever the R
+      CMD check failure was either "warning" or "error".
+
+  - Output that was formerly shown only on manual runs when the
+    "debug" checkbox was selected is now always shown.  The "debug"
+    option has been changed to "dry-run", which results in the debug
+    output being shown but the actually check, and those set-up steps
```
Comment (gsrohde) [inline NEWS.md:97]: actually --> actual

**Review by gsrohde — COMMENTED** [review]
I noticed one minor typo in the NEWS.md file.
---
I think this is _probably_ OK  now, but I do get one warning when I run `R CMD check --as-cran` on my machine, a warning that doesn't show up in the GitHub Workflows runs.

**Comment by eloch216** [comment]
> I think this is _probably_ OK now, but I do get one warning when I run `R CMD check --as-cran` on my machine, a warning that doesn't show up in the GitHub Workflows runs:
> 
> ```
> * checking sizes of PDF files under ‘inst/doc’ ... WARNING
>   ‘gs+qpdf’ made some significant size reductions:
>      compacted ‘a_practical_guide_to_biocro.pdf’ from 1119Kb to 692Kb
>      compacted ‘an_introduction_to_biocro.pdf’ from 744Kb to 253Kb
>   consider running tools::compactPDF(gs_quality = "ebook") on these files
> * checking installed files from ‘inst/doc’ ... OK
> ```

This isn't anything inherent to the package itself. You can make it go away if you specify `--compact-vignettes=both` when you build the package before checking it.

My understanding of the CRAN submission process is that we need to submit a tarball created by calling `R CMD build`. We can specify `--compact-vignettes=both` when doing that, and then there shouldn't be any vignette size issues.

**Comment by eloch216** [comment]
There was another harmonic oscillator test failure on one of the operating systems. This could be a potential issue when submitting to CRAN. If we are not lucky, they might encounter a test failure. Maybe we could set a seed number before running this test to try making it more reliable.

**Comment by eloch216** [comment]
> There was another harmonic oscillator test failure on one of the operating systems. This could be a potential issue when submitting to CRAN. If we are not lucky, they might encounter a test failure. Maybe we could set a seed number before running this test to try making it more reliable.

All the tests passed with a seed set (`set.seed(1234)`). I know this makes the test a bit less thorough. However, the fully random version of the test is not very helpful for troubleshooting since the failures cannot be repeated.

**Comment by gsrohde** [comment]
> All the tests passed with a seed set (`set.seed(1234)`). I know this makes the test a bit less thorough.

My initial thought was that we could compensate for not truly randomizing the tests somewhat by increasing number of trials.  But then I questioned how important this was.

If we think of this as a regression test, providing a backstop against the code changing in unexpected ways, then the fact that we use the same randomly-selected set of parameters and initial values each time the test is run doesn't seem so important.

But on the other hand, this test is fundamentally different from other tests that involve running a simulation because here, unlike the tests that model plant growth, we have a simple physics-based prediction of what the result of the simulation should be.  So we are also testing whether our simulator is working in accord with this physics-based model.  So this test is more than just a regression test.

I'm OK, though, at least for now, with using this seed-based approach to making the test less "flakey".  (For some perspective on flakey tests, and I've probably sent this link out before, see, for example, https://testing.googleblog.com/2016/05/flaky-tests-at-google-and-how-we.html.)  In the longer term, though, it might be better just to relax the tolerance level somewhat, or to relax it for certain solvers if it is particular solvers that are causing the issue.  That, however, could be part of a separate issue and/or PR.

**Pull request merged**